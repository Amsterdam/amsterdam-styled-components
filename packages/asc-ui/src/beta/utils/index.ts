import deprecatedWarning from './deprecatedWarning'

export {
  themeColor,
  srOnlyStyle,
  breakpoint,
  svgFill,
  fluidTypoStyle,
  typographyStyle,
  showAboveBackDrop,
  getValueFromTheme,
  themeSpacing,
} from './themeUtils'
export { valueFromObject } from './valueFromObject'
export { fromProps } from './fromProps'

export { deprecatedWarning }
