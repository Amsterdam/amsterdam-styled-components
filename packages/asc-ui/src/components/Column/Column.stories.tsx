import * as React from 'react'
import { storiesOf } from '@storybook/react'
import Column from './Column'


const ColumnStory: React.FC<{}> = () => (
  <Column debug span={{ large: 8, big: 6, medium: 5, small: 4 }}>
    <span>children go here</span>
  </Column>
)

storiesOf('Atoms/Column', module)
  .add('Column', () => <ColumnStory />)
