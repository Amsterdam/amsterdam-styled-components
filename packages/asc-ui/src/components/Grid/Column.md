# Grid

The grid components depend on a site's theme configuration and specifically the `layouts` section of the theme. A single layout consists of a set of props for the number of columns, the width of the gutters, the margin surrounding the grid and the `min-width` and `max-width` queries, like so:

```javascript
{
  "layouts": {
    "major": {
      "columns": 66, # (number) number of columns that should be available between the values of max and min; required
      "gutter": 20,  # (px) space between two columns; optional (default: 0)
      "margin": 42,  # (px) space on either side of the grid; optional (default: 0)
      "max": 2560,   # (px) max-width media query value; optional
      "min": 1820    # (px) min-width media query value; optional, but either min or max should be present
    },
    "minor": {
      "columns": 12,
      "gutter": 16,
      "margin": 21,
      "max": 1820
    }
  }
}
```

The theme configuration can contain any number of layouts.

Layout names are used to generate media queries for `<Column />` and `<Row />` components and don't adhere to any restrictions.


# `<Row />`

The `Row` component ensures that its children are positionable in a, well, row. It does so by applying `flex-box` style rules. By default, a `Row`'s children are spaced with space between them. The alignment behaviour can be changed by passing a value for `halign` and/or `valign` prop.
A `Row` component can take a `debug` prop. Setting this prop will render a label that shows the name of the layout as well as the min and max values of that layout. It will also render a linearly repeated background image that shows the position of the columns in the grid.

The below example will align the child `Column` component with the remaining space around them. Note that using a different alignmen than the default `space-between` will not ensure the alignment of the `Column` components to the grid.

```jsx
<Row halign="space-around">
  ...
</Row>
```

## Debugging

Setting the `debug` prop on a `<Row />` component will show both a repeating linear background images that indicated where the columns will be placed as well as as label that shows the current layout name based on the width of the screen. The `debugColor` prop can be given a valid color string (hex, hsl, rgb, rbga) which will change the color of the linear gradient.

```jsx
<Row debug debugColor="rgba(0,255,0,0.5)"> // 50% green
  ...
</Row>
```



# `<Column />`

## Spanning

A `Column` component can be configured to span a certain amount of columns at all layouts or an amount of columns per layout:

```jsx
// always span two columns, regardless of layout
<Column span={2}>...</Column>

// conditionally span columns
<Column span={{ major: 4, minor: 3 }}>...</Column>
```

A column can also be pushed a number of columns to the right as can it be given an order the same way it can be given a span:

## Ordering

```jsx
// set the order regardless of layout
<Column order={2}>...</Column>
<Column order={1}>...</Column>

// set a specific order per layout
<Column order={{ major: 1, minor: 2 }}>...</Column>
<Column order={{ major: 2, minor: 1 }}>...</Column> // in minor, this column will be the first
```

## Pushing

```jsx
// always push two columns to the right
<Column push={2}>...</Column>


// conditionally push columns
<Column push={{ major: 0, minor: 1 }}>...</Column>
```

## Debugging

A `Column` component can also be passed a `debug` prop. Setting this prop to 'true' will render a label that shows the current span of that column. As the layout changes, the label on the column will also change value.

```jsx
<Column debug span={2}>...</Column>
```
