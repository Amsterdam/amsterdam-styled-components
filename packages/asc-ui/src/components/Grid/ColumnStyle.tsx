import React, { ReactNode } from 'react'
import styled, { css } from 'styled-components'
import {
  mediaQuery,
  colCount,
  spanWidth,
  pushWidth,
  margin,
} from '../../utils/grid'
import { TypeFlexPosition } from '.'
import { Theme } from '../../types'

export type TypeProps = {
  alignSelf?: TypeFlexPosition
  children: ReactNode
  className?: string
  debug?: boolean
  debugColor?: string
  id?: string
  order?: Theme.TypeSpan
  parentSpan?: Theme.TypeSpan
  push?: Theme.TypeSpan
  span: Theme.TypeSpan
  wrap?: boolean
}

const ColumnStyle = styled(({ className, children, id, ...rest }) => {
  const dataProps = {}
  Object.keys(rest)
    .filter((key) => key.startsWith('data-') || key.startsWith('aria-'))
    .forEach((key) => {
      dataProps[key] = rest[key]
    })

  return (
    <div className={className} id={id} {...dataProps}>
      {children}
    </div>
  )
})`
  box-sizing: border-box;
  width: 100%;
  display: flex;
  align-self: ${({ alignSelf }: { alignSelf?: string }) => alignSelf};
  justify-content: space-between;

  ${({ wrap }: { wrap?: boolean }) =>
    wrap &&
    css`
      flex-wrap: wrap;
    `}
  ${({ order }: { order?: number }) =>
    order && typeof order !== 'number'
      ? Object.keys(order).map(
          (layoutId) => css`
            @media ${mediaQuery(layoutId)} {
              order: ${order[layoutId]};
            }
          `,
        )
      : css`
          order: ${order};
        `}
  ${({
    debug,
    span,
    parentSpan,
    push = 0,
    theme,
  }: {
    debug?: boolean
    parentSpan?: Theme.TypeSpan
    span: Theme.TypeSpan
    push?: Theme.TypeSpan
    theme: Theme.ThemeInterface
  }) =>
    Object.keys(theme.layouts).map((layoutId) => {
      const spanCount = colCount(span, layoutId)
      const pushCount = colCount(push, layoutId)

      if (spanCount < 1 || pushCount < 0) {
        return null
      }

      return css`
        @media ${mediaQuery(layoutId)} {
          ${debug &&
          css`
            &::before {
              content: 'span ${spanCount}';
            }
          `}
          ${pushCount > 0 &&
          margin(layoutId)({ theme }) > 0 &&
          css`
            margin-left: ${pushWidth({ layoutId, push, parentSpan })};
          `}
          max-width: ${spanWidth({ layoutId, span, parentSpan })};
        }
      `
    })};
  ${({ debug, debugColor }: { debug?: boolean; debugColor?: string }) =>
    debug &&
    css`
      position: relative;
      border: 1px solid;
      border-color: ${debugColor};

      &::before {
        background: white;
        position: absolute;
        top: -12px;
        left: 4px;
        padding: 2px 4px;
        border: 1px solid ${debugColor};
        font: 12px sans-serif;
      }
    `}
`

ColumnStyle.defaultProps = {
  order: 0,
}

export default ColumnStyle
