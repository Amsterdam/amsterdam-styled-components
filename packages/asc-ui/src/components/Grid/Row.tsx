import type { TypeProps } from './RowStyle'
import RowStyle from './RowStyle'

function Row({ children, debug, ...props }: TypeProps) {
  return (
    <RowStyle {...props} debug={debug}>
      {debug && (
        <span className="layout-label">
          <span />
        </span>
      )}
      {children}
    </RowStyle>
  )
}

Row.defaultProps = {
  debug: false,
  debugColor: 'rgba(51,102,153,0.75)',
  hasMargin: true,
}

export default Row
