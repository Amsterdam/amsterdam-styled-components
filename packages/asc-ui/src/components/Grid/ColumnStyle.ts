import React from 'react'
import styled, { Theme, css, ascDefaultTheme } from '@datapunt/asc-core'
import { mediaQuery, colCount, spanWidth, pushWidth } from '../../utils/grid'
import { TypeFlexPosition } from '.'

const { layouts } = ascDefaultTheme

export type TypeProps = {
  alignSelf?: TypeFlexPosition
  children: React.ReactNode
  className?: string
  debug?: boolean
  debugColor?: string
  id?: Theme.TypeLayout
  order?: Theme.TypeSpan
  parentSpan?: Theme.TypeSpan
  push?: Theme.TypeSpan
  span: Theme.TypeSpan
}

const ColumnStyle = styled.div<TypeProps>`
  box-sizing: border-box;
  width: 100%;
  display: flex;
  align-self: ${({ alignSelf }) => alignSelf};
  justify-content: space-between;
  ${({ order }) =>
    order && typeof order !== 'number'
      ? Object.keys(order).map(id => {
          const layoutId = id as Theme.TypeLayout
          return css`
            @media ${mediaQuery(layoutId)} {
              order: ${order[layoutId]};
            }
          `
        })
      : css`
          order: ${order};
        `}
  ${({
    debug,
    span,
    parentSpan,
    push = 0,
  }: {
    debug?: boolean
    parentSpan?: Theme.TypeSpan
    span: Theme.TypeSpan
    push?: Theme.TypeSpan
  }) =>
    Object.keys(layouts).map(id => {
      const layoutId = id as Theme.TypeLayout
      const spanCount = colCount(span, layoutId)
      const pushCount = colCount(push, layoutId)
      if (spanCount < 1 || pushCount < 0) {
        return null
      }
      return css`
        @media ${mediaQuery(layoutId)} {
          ${debug &&
            css`
            &:before {
              content: 'span ${spanCount}';
            }
          `}
          max-width: ${spanWidth({ layoutId, span, parentSpan })};
          ${pushCount > 0 &&
            css`
              margin-left: ${pushWidth({ layoutId, push, parentSpan })};
            `}
        }
      `
    })};
  ${({ debug, debugColor }) =>
    debug &&
    css`
      position: relative;
      border: 1px solid;
      border-color: ${debugColor};

      &:before {
        background: white;
        position: absolute;
        top: -12px;
        left: 4px;
        padding: 2px 4px;
        border: 1px solid ${debugColor};
        font: 12px sans-serif;
      }
    `}
`

ColumnStyle.defaultProps = {
  order: 0,
}

export default ColumnStyle
