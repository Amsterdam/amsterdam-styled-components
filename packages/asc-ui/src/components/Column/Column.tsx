import React from 'react'
import { Color } from 'csstype'
import ColumnStyle from '../../styles/components/ColumnStyle'
import { Theme } from '../../styles'

type Props = {
  children: React.ReactNode
  debug?: boolean
  debugColor?: Color
  id?: keyof Theme.GridInterface
  order?: number
  push?: Theme.TypeSpan
  span: Theme.TypeSpan
}

type Context = {
  parentSpan?: Theme.TypeSpan
}

const ParentContext = React.createContext<Context>({})

const Column: React.FC<Props> = ({ children, span, id, ...props }) => {
  const { parentSpan } = React.useContext(ParentContext)

  return (
    <ColumnStyle parentSpan={parentSpan} span={span} id={id} {...props}>
      {children}
    </ColumnStyle>
  )
}

Column.defaultProps = {
  debug: false,
  debugColor: 'rgba(51,102,153,1)',
  order: 0,
  push: 0,
  span: 1,
}

export default Column
