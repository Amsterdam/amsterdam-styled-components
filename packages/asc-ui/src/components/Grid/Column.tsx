import type { FunctionComponent, ReactNode } from 'react'
import { createContext, useContext } from 'react'
import type { TypeProps as ColumnStyleProps } from './ColumnStyle'
import ColumnStyle from './ColumnStyle'
import type { Theme } from '../../types'

type Props = {
  children: ReactNode
  className?: string
  debug?: boolean
  debugColor?: string
  id?: string
  order?: Theme.TypeSpan
  push?: Theme.TypeSpan
  span: Theme.TypeSpan
  wrap?: boolean
} & ColumnStyleProps

interface Context {
  parentSpan?: Theme.TypeSpan
}

const ParentContext = createContext<Context>({})

const Column: FunctionComponent<Props> = ({ children, span, ...props }) => {
  const { parentSpan } = useContext<Context>(ParentContext)

  return (
    <ParentContext.Provider value={{ parentSpan: span }}>
      <ColumnStyle parentSpan={parentSpan} span={span} {...props}>
        {children}
      </ColumnStyle>
    </ParentContext.Provider>
  )
}

Column.defaultProps = {
  debug: false,
  debugColor: 'rgba(51,102,153,0.5)',
  order: 0,
  push: 0,
  span: 1,
  wrap: false,
}

export default Column
