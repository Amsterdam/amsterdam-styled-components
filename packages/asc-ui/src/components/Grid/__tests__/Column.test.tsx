import React from 'react'
import { render, cleanup } from '@testing-library/react'
import 'jest-styled-components'
import { ascDefaultTheme } from '@datapunt/asc-core'
import { mount } from 'enzyme'
import Row from '../Row'
import Column from '../Column'
import ColumnStyle from '../ColumnStyle'
import { ThemeProvider } from '../../../index'
import { mediaQuery } from '../../../utils/grid'

const mq = (layoutId: string) =>
  mediaQuery(layoutId)({ theme: ascDefaultTheme })

describe('Column', () => {
  it('should filter invalid attributes', () => {
    render(
      <ThemeProvider>
        <Row>
          <Column data-testid="foo" aria-labelledby="bar" wrap span={1}>
            <span>Child right here</span>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(document.querySelectorAll('[data-testid="foo"]')).toHaveLength(1)
    expect(document.querySelectorAll('[aria-labelledby="bar"]')).toHaveLength(1)
    expect(document.querySelectorAll('[wrap]')).toHaveLength(0)
  })
})

describe('ColumnStyle', () => {
  afterEach(cleanup)

  it('should set parentSpan for nested column components', () => {
    const tree = mount(
      <ThemeProvider>
        <Row>
          <Column span={10}>
            <Column data-testid="span5" span={5}>
              Child right here
            </Column>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(
      tree
        .find(ColumnStyle)
        .first()
        .props().parentSpan,
    ).toBeUndefined()
    expect(
      tree
        .find(ColumnStyle)
        .last()
        .props().parentSpan,
    ).toEqual(10)
  })

  it('should show debug information', () => {
    const span = 7
    const { rerender, queryByTestId } = render(
      <ThemeProvider>
        <Row>
          <Column data-testid="col" span={span}>
            <span>Child right here</span>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(queryByTestId('col')).not.toHaveStyleRule(
      'content',
      `span ${span}`,
      { modifier: '::before' },
    )

    rerender(
      <ThemeProvider>
        <Row>
          <Column data-testid="col" span={span} debug>
            <span>Another child</span>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(queryByTestId('col')).toHaveStyleRule('content', `'span ${span}'`, {
      modifier: '::before',
      media: mq('large'),
    })
    expect(queryByTestId('col')).toHaveStyleRule('content', `'span ${span}'`, {
      modifier: '::before',
      media: mq('big'),
    })
    expect(queryByTestId('col')).toHaveStyleRule('content', `'span ${span}'`, {
      modifier: '::before',
      media: mq('medium'),
    })
    expect(queryByTestId('col')).toHaveStyleRule('content', `'span ${span}'`, {
      modifier: '::before',
      media: mq('small'),
    })
  })

  it('should set order values', () => {
    const { queryByTestId, rerender } = render(
      <ThemeProvider>
        <Row>
          <Column data-testid="col" order={2} span={4}>
            <span>Another child</span>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(queryByTestId('col')).toHaveStyleRule('order', '2')

    rerender(
      <ThemeProvider>
        <Row>
          <Column
            data-testid="col"
            order={{ large: 8, big: 6, medium: 4, small: 2 }}
            span={4}
          >
            <span>Another child</span>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(queryByTestId('col')).toHaveStyleRule('order', '8', {
      media: mq('large'),
    })
    expect(queryByTestId('col')).toHaveStyleRule('order', '6', {
      media: mq('big'),
    })
    expect(queryByTestId('col')).toHaveStyleRule('order', '4', {
      media: mq('medium'),
    })
    expect(queryByTestId('col')).toHaveStyleRule('order', '2', {
      media: mq('small'),
    })
  })

  it('should set push values', () => {
    const { queryByTestId, rerender } = render(
      <ThemeProvider>
        <Row>
          <Column data-testid="span4" span={4} push={2}>
            <span>Another child</span>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(queryByTestId('span4')).toHaveStyleRule(
      'margin-left',
      expect.stringContaining('* 2'),
      {
        media: mq('large'),
      },
    )

    expect(queryByTestId('span4')).toHaveStyleRule(
      'margin-left',
      expect.stringContaining('* 2'),
      {
        media: mq('big'),
      },
    )

    expect(queryByTestId('span4')).toHaveStyleRule(
      'margin-left',
      expect.stringContaining('* 2'),
      {
        media: mq('medium'),
      },
    )

    expect(queryByTestId('span4')).toHaveStyleRule(
      'margin-left',
      expect.stringContaining('* 2'),
      {
        media: mq('small'),
      },
    )

    rerender(
      <ThemeProvider>
        <Row>
          <Column
            data-testid="span4"
            span={4}
            push={{ large: 8, big: 6, medium: 4, small: 2 }}
          >
            <span>Another child</span>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(queryByTestId('span4')).toHaveStyleRule(
      'margin-left',
      expect.stringContaining('* 8'),
      {
        media: mq('large'),
      },
    )

    expect(queryByTestId('span4')).toHaveStyleRule(
      'margin-left',
      expect.stringContaining('* 6'),
      {
        media: mq('big'),
      },
    )

    expect(queryByTestId('span4')).toHaveStyleRule(
      'margin-left',
      expect.stringContaining('* 4'),
      {
        media: mq('medium'),
      },
    )

    expect(queryByTestId('span4')).toHaveStyleRule(
      'margin-left',
      expect.stringContaining('* 2'),
      {
        media: mq('small'),
      },
    )
  })

  it('should wrap its contents', () => {
    const { queryByTestId } = render(
      <ThemeProvider>
        <Row>
          <Column wrap span={10} data-testid="span10_wrap">
            <Column data-testid="span5" span={5}>
              Child right here
            </Column>

            <Column span={10}>Child right here</Column>

            <Column span={0}>Child right here</Column>
          </Column>
        </Row>
      </ThemeProvider>,
    )

    expect(queryByTestId('span10_wrap')).toHaveStyleRule('flex-wrap', 'wrap')
    expect(queryByTestId('span5')).not.toHaveStyleRule('flex-wrap', 'wrap')
  })
})
