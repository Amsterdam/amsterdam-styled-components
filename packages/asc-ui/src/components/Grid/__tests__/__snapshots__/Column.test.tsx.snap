// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnStyle should set push values 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  max-width: 1920px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 .layout-label {
  position: absolute;
  top: -20px;
  width: 100%;
}

.c0 .layout-label::before {
  position: absolute;
  padding: 2px 4px;
  font: 12px sans-serif;
  background: white;
  height: 20px;
}

.c1 {
  box-sizing: border-box;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

@media (min-width:2000px) {
  .c0 {
    padding-left: 60px;
    padding-right: 60px;
  }
}

@media (min-width:1000px) and (max-width:1999px) {
  .c0 {
    padding-left: 40px;
    padding-right: 40px;
  }
}

@media (min-width:500px) and (max-width:999px) {
  .c0 {
    padding-left: 20px;
    padding-right: 20px;
  }
}

@media (min-width:2000px) {
  .c1 {
    margin-left: calc((((100% - 578px) / 18) * 2) + 34px + 34px);
    max-width: calc((((100% - 578px) / 18) * 4) + 102px + 0px);
  }
}

@media (min-width:1000px) and (max-width:1999px) {
  .c1 {
    margin-left: calc((((100% - 264px) / 12) * 2) + 24px + 24px);
    max-width: calc((((100% - 264px) / 12) * 4) + 72px + 0px);
  }
}

@media (min-width:500px) and (max-width:999px) {
  .c1 {
    margin-left: calc((((100% - 140px) / 8) * 2) + 20px + 20px);
    max-width: calc((((100% - 140px) / 8) * 4) + 60px + 0px);
  }
}

@media (max-width:499px) {
  .c1 {
    max-width: 100%;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="span4"
  >
    <span>
      Another child
    </span>
  </div>
</div>
`;
